File: php/src/Chat/MessagesService.php
<?php

declare(strict_types=1);

namespace Buble\Chat;

use Buble\Http\TransportInterface;
use Buble\Streaming\SseEvent;
use Buble\Streaming\StreamText;

final class MessagesService
{
    public function __construct(private readonly TransportInterface $transport)
    {
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    public function create(array $body): array
    {
        $body['stream'] = false;
        return $this->transport->request('POST', '/api/v1/messages', $body);
    }

    /**
     * @param array<string, mixed> $body
     * @return \Generator<int, SseEvent>
     */
    public function stream(array $body): \Generator
    {
        $body['stream'] = true;
        yield from $this->transport->stream('/api/v1/messages', $body);
    }

    /**
     * @param array<string, mixed> $body
     * @return \Generator<int, string>
     */
    public function streamText(array $body): \Generator
    {
        yield from StreamText::fromEvents($this->stream($body), 'anthropic');
    }
}

File: dotnet/src/Buble.Sdk/Chat/MessagesService.cs
using Buble.Sdk.Http;
using Buble.Sdk.Streaming;
using System.Text.Json.Nodes;

namespace Buble.Sdk.Chat;

public sealed class MessagesService
{
    private readonly BubleHttpClient _http;

    internal MessagesService(BubleHttpClient http)
    {
        _http = http;
    }

    public Task<JsonObject?> CreateAsync(
        JsonObject body,
        CancellationToken cancellationToken = default)
    {
        var payload = JsonPayload.Copy(body);
        payload["stream"] = false;
        return _http.PostAsync<JsonObject>("/api/v1/messages", payload, cancellationToken: cancellationToken);
    }

    public Task<JsonObject?> CreateAsync(
        IDictionary<string, object?> body,
        CancellationToken cancellationToken = default)
    {
        return CreateAsync(JsonPayload.FromDictionary(body), cancellationToken);
    }

    public IAsyncEnumerable<ServerSentEvent> StreamAsync(
        JsonObject body,
        CancellationToken cancellationToken = default)
    {
        var payload = JsonPayload.Copy(body);
        payload["stream"] = true;
        return ServerSentEventParser.ParseAsync(
            _http.StreamLinesAsync("/api/v1/messages", payload, cancellationToken: cancellationToken),
            cancellationToken);
    }

    public IAsyncEnumerable<string> StreamTextAsync(
        JsonObject body,
        CancellationToken cancellationToken = default)
    {
        return StreamText.FromEventsAsync(StreamAsync(body, cancellationToken), StreamProtocol.Anthropic, cancellationToken);
    }

    public IAsyncEnumerable<ServerSentEvent> StreamAsync(
        IDictionary<string, object?> body,
        CancellationToken cancellationToken = default)
    {
        return StreamAsync(JsonPayload.FromDictionary(body), cancellationToken);
    }

    public IAsyncEnumerable<string> StreamTextAsync(
        IDictionary<string, object?> body,
        CancellationToken cancellationToken = default)
    {
        return StreamTextAsync(JsonPayload.FromDictionary(body), cancellationToken);
    }
}
